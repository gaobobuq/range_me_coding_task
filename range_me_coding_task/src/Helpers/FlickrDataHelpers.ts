class FlickrDataHelpers {
  public static extractAuthorName = (authorString?: string): string => {
    return !!authorString
      ? authorString.substring(authorString.indexOf('("') + 2, authorString.lastIndexOf('")')).trim()
      : '';
  };

  public static extractAuthorEmail = (authorString?: string): string => {
    return !!authorString
      ? authorString.substring(0, authorString.lastIndexOf('("')).trim()
      : '';
  };
}

export default FlickrDataHelpers;
