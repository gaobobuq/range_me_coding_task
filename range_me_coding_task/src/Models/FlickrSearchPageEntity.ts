import { JsonFlickrFeedEntity, QueryParametersEntity } from '.';

export default class FlickrSearchPageEntity {
  public queryParameters?: QueryParametersEntity;
  public flickrFeed?: JsonFlickrFeedEntity;
}
