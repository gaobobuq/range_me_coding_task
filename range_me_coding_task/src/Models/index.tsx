export { default as FlickrSearchPageEntity } from './FlickrSearchPageEntity';
export { default as QueryParametersEntity } from './QueryParametersEntity';
export { default as MediaEntity } from './MediaEntity';
export { default as ItemsEntity } from './ItemsEntity';
export { default as JsonFlickrFeedEntity } from './JsonFlickrFeedEntity';
