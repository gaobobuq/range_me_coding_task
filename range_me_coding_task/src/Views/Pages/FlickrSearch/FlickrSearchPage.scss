
.flickr-search-page {
  margin: auto;
  max-width: 1000px;

  .search-box {
    margin: 20px;

    .search-icon {
      width: 35px;
      vertical-align: bottom;
      padding: 1px;
      margin: 0 10px;
    }

    .input-search-keyword {
      border-top: none;
      border-left: none;
      border-right: none;
      border-color: gray;
      padding: 10px;
      font-size: 20px;
      width: 350px;
    }
  }

  .loading-result {
    margin: auto;
    margin-top: 100px;
    width: 25%;

    .loading-text {
      font-size: 23px;
      margin-top: 10px;
    }
  }

  .no-result-found {
    margin: auto;
    width: 70%;

    img {
      width: 300px;
    }
  }

  .flickr-feed-items{
    margin-top: 10px;
    margin-bottom: 50px;
  }
}
