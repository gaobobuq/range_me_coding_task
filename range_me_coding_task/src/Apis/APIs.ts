import flickrSampleData from '../MockedData/FlickrSampleData';
import { JsonFlickrFeedEntity } from '../Models';

/**
 * PaginatedFlickrData interface
 */
export interface PaginatedFlickrData {
  pageCount: number;
  currentPageIndex: number;
  jsonFlickrFeedEntity: JsonFlickrFeedEntity;
}

/**
 * APIs class
 *
 * api collection of the application
 */
export default class APIs {
  /**
   * fetch flickr data method
   * get request to fetch flickr data
   * @param searchKeyword {string} the input searchKeyword text
   * @param currentPageIndex {number} current page index
   * @param itemCountPerPage {number} number of items displayed per page
   */
  static fetchFlickrData = (searchKeyword: string, currentPageIndex: number, itemCountPerPage: number): PaginatedFlickrData => {
    const filteredItems = flickrSampleData.items?.filter(e => {
      return e.author?.includes(searchKeyword)
            || e.tags?.includes(searchKeyword)
            || (
              e.date_taken && new Date(e.date_taken).toUTCString().includes(searchKeyword)
            );
    });

    const items = filteredItems ? filteredItems.slice((currentPageIndex - 1) * itemCountPerPage, currentPageIndex * itemCountPerPage) : [];
    const parsedJsonFlickrFeedEntity = new JsonFlickrFeedEntity({
      ...flickrSampleData,
      items,
    });

    return {
      currentPageIndex: currentPageIndex,
      pageCount: filteredItems ? Math.ceil(filteredItems.length / itemCountPerPage) : 1,
      jsonFlickrFeedEntity: parsedJsonFlickrFeedEntity,
    };
  };
}
