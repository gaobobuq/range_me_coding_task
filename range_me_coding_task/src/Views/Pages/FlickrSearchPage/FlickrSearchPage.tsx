import * as React from 'react';
import { useEffect, useState } from 'react';
import { Col, Row } from 'react-grid-system';
import flickrSampleData from '../../../MockedData/FlickrSampleData';
import { ItemsEntity, JsonFlickrFeedEntity } from '../../../Models';
import FlickrItemComponent from '../../Components/FlickrItem/FlickrItemComponent';
import '../../frontend.scss';

function FlickrSearchPage() {
  const [flickrFeedData, setFlickrFeedData] = useState<JsonFlickrFeedEntity>();

  // search flickr data event
  const fetchFlickrData = (keyword: string) => {
    const parsedJsonFlickrFeedEntity = new JsonFlickrFeedEntity(flickrSampleData);

    setFlickrFeedData(parsedJsonFlickrFeedEntity);
  };

  useEffect(() => {
    fetchFlickrData('');
  }, []);

  return (
    <div className="flickr-search-page">
      <Row>
        <Col md={12} sm={12}>
          {flickrFeedData && flickrFeedData.items?.length > 0 ? (
            <div data-testid="flickr-data-items-wrapper" className="flickr-feed-items">
              {flickrFeedData?.items?.map((item: ItemsEntity, index: number) => (
                <FlickrItemComponent key={item.title ?? `${index}`} {...item} />
              ))}
            </div>
          ) : (
            <div className="no-result-found">
              <img
                data-testid="no-result-found"
                src="no-result-found.png"
                alt="no results"
              />
            </div>
          )}
        </Col>
      </Row>
    </div>
  );
}

export default FlickrSearchPage;
