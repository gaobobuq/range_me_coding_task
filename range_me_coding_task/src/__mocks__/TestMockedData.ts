import { ItemsRawData, JsonFlickrFeedRawData } from '../Models/RawData';

export const itemsEntityRawData: ItemsRawData = {
  title: 'IMG_2481',
  link: 'https://www.flickr.com/photos/mickythepixel/51893643867/',
  media: {
    m: 'https://live.staticflickr.com/65535/51893643867_051d38f43e_m.jpg',
  },
  date_taken: '2022-02-19T12:08:25-08:00',
  description:
      ' <p><a href="https://www.flickr.com/people/mickythepixel/">micky the pixel</a> posted a photo:</p> <p><a href="https://www.flickr.com/photos/mickythepixel/51893643867/" title="IMG_2481"><img src="https://live.staticflickr.com/65535/51893643867_051d38f43e_m.jpg" width="230" height="240" alt="IMG_2481" /></a></p> ',
  published: '2022-02-21T09:23:45Z',
  author: 'nobody@flickr.com ("micky the pixel")',
  author_id: '7141646@N02',
  tags: 'buch book livre zeitschrift magazine verlagernstkeilsnachfolger diegartenlaubeâ€“illustriertesfamilienblatt diegartenlaube illustration wilhelmvonkaulbach reinekefuchs hinrichtung execution',
};

export const jsonFlickrFeedRawDataOneInstance: JsonFlickrFeedRawData = {
  title: 'Uploads from everyone',
  link: 'https://www.flickr.com/photos/',
  description: '',
  modified: '2022-02-21T09:23:41Z',
  generator: 'https://www.flickr.com',
  items: [
    {
      title: ' ',
      link: 'https://www.flickr.com/photos/192801292@N06/51893643787/',
      media: {
        m: 'https://live.staticflickr.com/65535/51893643787_c2269ca5d9_m.jpg',
      },
      date_taken: '2022-02-21T14:27:58-08:00',
      description:
                ' <p><a href="https://www.flickr.com/people/192801292@N06/">abdulrazacksa</a> posted a photo:</p> <p><a href="https://www.flickr.com/photos/192801292@N06/51893643787/" title=" "><img src="https://live.staticflickr.com/65535/51893643787_c2269ca5d9_m.jpg" width="240" height="195" alt=" " /></a></p> ',
      published: '2022-02-21T09:23:41Z',
      author: 'nobody@flickr.com ("abdulrazacksa")',
      author_id: '192801292@N06',
      tags: '',
    },
  ],
};

export const jsonFlickrFeedRawDataNoItems: JsonFlickrFeedRawData = {
  title: 'Uploads from everyone',
  link: 'https://www.flickr.com/photos/',
  description: '',
  modified: '2022-02-21T09:23:41Z',
  generator: 'https://www.flickr.com',
  items: [],
};
