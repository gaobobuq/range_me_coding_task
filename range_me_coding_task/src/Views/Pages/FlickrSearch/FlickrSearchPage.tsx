import { CircularProgress, Pagination } from '@mui/material';
// @ts-ignore
import debounce from 'lodash.debounce';
import * as React from 'react';
import { useCallback, useEffect, useState } from 'react';
import { Col, Row } from 'react-grid-system';
import APIs from '../../../Apis/APIs';
import { ITEM_COUNT_PER_PAGE } from '../../../Constants';
import { ItemsEntity, JsonFlickrFeedEntity } from '../../../Models';
import FlickrItemComponent from '../../Components/FlickrItem/FlickrItemComponent';

import '../../frontend.scss';
import './FlickrSearchPage.scss';

/**
 * FlickrSearchPage Component
 * Page to search and display Flickr items
 */
function FlickrSearchPage() {
  const [searchKeyword, setSearchKeyword] = useState<string>('');
  const [flickrFeedData, setFlickrFeedData] = useState<JsonFlickrFeedEntity>();
  const [currentPageIndex, setCurrentPageIndex] = useState<number>(1);
  const [pageCount, setPageCount] = useState<number>(1);
  const [isLoadingData, setIsLoadingData] = useState<boolean>(true);

  // search flickr data event
  const fetchFlickrData = (keyword: string) => {
    const paginatedFlickrData = APIs.fetchFlickrData(keyword, currentPageIndex, ITEM_COUNT_PER_PAGE);
    setFlickrFeedData(paginatedFlickrData.jsonFlickrFeedEntity);
    setPageCount(paginatedFlickrData.pageCount);
    setIsLoadingData(false);
  };

  useEffect(() => {
    fetchFlickrData('');
  }, [currentPageIndex]);

  const debouncedSearch = useCallback(
    debounce((newSearchKeyword: string) => fetchFlickrData(newSearchKeyword), 1000),
    [],
  );

  const searchKeywordChanged = (newSearchKeyword: string) => {
    setSearchKeyword(newSearchKeyword);
    setCurrentPageIndex(1);
    setIsLoadingData(true);
    debouncedSearch(newSearchKeyword);
  };

  return (
    <div className="flickr-search-page">
      <Row>
        {/* search keyword input */}

        <Col md={12} sm={12}>
          <div className="search-box">
            <img className="search-icon" src="/search-icon.svg" alt="search-icon" />
            <input
              data-testid="input-search-keyword"
              className="input-search-keyword"
              placeholder="Search by author, tag or date"
              value={searchKeyword}
              onChange={event => {
                searchKeywordChanged(event.target.value);
              }}
            />
          </div>
        </Col>

        {/* display search result items */}
        {
          isLoadingData
            ? (
              <div data-testid="loading-result" className="loading-result">
                <CircularProgress size={80} />
                <div className="loading-text">Loading</div>
              </div>
            )
            : (
              <Col md={12} sm={12}>
                {flickrFeedData?.items?.length > 0 ? (
                  <div data-testid="flickr-data-items-wrapper" className="flickr-feed-items">
                    {flickrFeedData?.items?.map((item: ItemsEntity, index: number) => (
                      <FlickrItemComponent key={item.title ?? `${index}`} {...item} />
                    ))}

                    {
                        pageCount > 1
                        && (
                        <Pagination
                          data-testid="pagination"
                          className="pagination"
                          page={currentPageIndex}
                          count={pageCount}
                          variant="outlined"
                          shape="rounded"
                          onChange={(event: React.ChangeEvent<unknown>, page: number) => {
                            setCurrentPageIndex(page);
                          }}
                        />
                        )
                      }
                  </div>
                ) : (
                  <div data-testid="no-result-found" className="no-result-found">
                    <img
                      src="no-result-found.png"
                      alt="no results"
                    />
                  </div>
                )}
              </Col>
            )
        }
      </Row>
    </div>
  );
}

export default FlickrSearchPage;
