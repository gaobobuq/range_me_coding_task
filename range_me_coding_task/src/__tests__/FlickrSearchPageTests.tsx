import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import APIs, { PaginatedFlickrData } from '../Apis/APIs';
import FlickrDataHelpers from '../Helpers/FlickrDataHelpers';
import flickrSampleData from '../MockedData/FlickrSampleData';
import { JsonFlickrFeedEntity } from '../Models';
import FlickrSearchPage from '../Views/Pages/FlickrSearch/FlickrSearchPage';
import { jsonFlickrFeedRawDataNoItems, jsonFlickrFeedRawDataOneInstance } from '../__mocks__/TestMockedData';

jest.mock('../Apis/APIs.ts');

/**
 * API mockups
 */
const mockFetchFlickrDataNoPagination = () => {
  (APIs.fetchFlickrData as jest.Mock).mockReturnValue({
    currentPageIndex: 1,
    pageCount: 1,
    jsonFlickrFeedEntity: new JsonFlickrFeedEntity(jsonFlickrFeedRawDataOneInstance),
  } as PaginatedFlickrData);
};

const mockFetchFlickrDataHasPagination = () => {
  (APIs.fetchFlickrData as jest.Mock).mockReturnValue({
    currentPageIndex: 1,
    pageCount: 2,
    jsonFlickrFeedEntity: new JsonFlickrFeedEntity(flickrSampleData),
  } as PaginatedFlickrData);
};

const mockFetchFlickrDataHasNoItems = () => {
  (APIs.fetchFlickrData as jest.Mock).mockReturnValue({
    currentPageIndex: 0,
    pageCount: 0,
    jsonFlickrFeedEntity: new JsonFlickrFeedEntity(jsonFlickrFeedRawDataNoItems),
  } as PaginatedFlickrData);
};

const mockFetchFlickrDataWith20Items = () => {
  (APIs.fetchFlickrData as jest.Mock).mockReturnValue({
    currentPageIndex: 1,
    pageCount: 4,
    jsonFlickrFeedEntity: new JsonFlickrFeedEntity(flickrSampleData),
  } as PaginatedFlickrData);
};

/**
 * Flickr Search Page Tests
 */
describe('(Tags: RMCT-004 FlickrSearchPage) Flickr Search Page Tests', () => {
  /**
   * Test scenarios
   */
  it('Flickr search page can be rendered', () => {
    mockFetchFlickrDataHasPagination();

    const wrapper = mount(
      <FlickrSearchPage />,
    );

    expect(wrapper).not.toBeNull();
  });

  it('One page should not have pagination', () => {
    mockFetchFlickrDataNoPagination();

    const wrapper = mount(
      <FlickrSearchPage />,
    );

    const paginationComponent = wrapper.find('*[data-testid="pagination"]');
    expect(paginationComponent.length).toBe(0);
  });

  it('Data items can be displayed on the page', () => {
    mockFetchFlickrDataWith20Items();

    const wrapper = mount(
      <FlickrSearchPage />,
    );

    const elements = wrapper.find('div[data-testid="author-name"]');

    // get mocked data items
    const jsonFlickrFeed = new JsonFlickrFeedEntity(flickrSampleData);

    // for each data item, we check whether the author's name is displayed on the page by index order
    elements.forEach((element: ReactWrapper, index: number) => {
      expect(element.text()).toBe(FlickrDataHelpers.extractAuthorName(jsonFlickrFeed.items[index].author));
    });
  });

  it('More than one pages should have pagination', () => {
    mockFetchFlickrDataHasPagination();

    const wrapper = mount(
      <FlickrSearchPage />,
    );

    // need to use timer to check after a short time, since initial isLoading state is true
    setTimeout(() => {
      wrapper.update();
      const paginationComponent = wrapper.find('*[data-testid="pagination"]');
      expect(paginationComponent.length).not.toBe(0);
    }, 100);
  });

  it('Search result has no items', () => {
    mockFetchFlickrDataHasNoItems();

    const wrapper = mount(
      <FlickrSearchPage />,
    );

    // need to use timer to check after a short time, since initial isLoading state is true
    setTimeout(() => {
      wrapper.update();
      const noResultFoundComponent = wrapper.find('*[data-testid="no-result-found"]');
      expect(noResultFoundComponent.length).toBe(1);
    }, 100);
  });
});
