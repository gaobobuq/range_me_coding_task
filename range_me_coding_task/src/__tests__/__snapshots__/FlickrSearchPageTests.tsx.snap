// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Tags: RCT-001) Flickr Search Page Tests Check Flickr search page details using snapshot 1`] = `
<div
  className="flickr-search-page"
>
  <div
    style={
      Object {
        "alignItems": "normal",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "flexShrink": 0,
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": -15,
        "marginRight": -15,
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "flexBasis": "100%",
          "flexGrow": 0,
          "flexShrink": 0,
          "left": "auto",
          "marginLeft": "0%",
          "maxWidth": "100%",
          "minHeight": 1,
          "paddingLeft": 15,
          "paddingRight": 15,
          "position": "relative",
          "right": "auto",
          "width": "100%",
        }
      }
    >
      <div
        className="search-box"
      >
        <img
          alt="search-icon"
          className="search-icon"
          src="/search-icon.svg"
        />
        <input
          className="input-search-keyword"
          data-testid="input-search-keyword"
          onChange={[Function]}
          placeholder="Search by author, tag or date"
          value=""
        />
      </div>
    </div>
    <div
      className="loading-result"
      data-testid="loading-result"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
      <div
        className="loading-text"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;

exports[`(Tags: RMCT-004 FlickrSearchPage) Flickr Search Page Tests Check Flickr search page details using snapshot 1`] = `
<div
  className="flickr-search-page"
>
  <div
    style={
      Object {
        "alignItems": "normal",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "flexShrink": 0,
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": -15,
        "marginRight": -15,
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "flexBasis": "100%",
          "flexGrow": 0,
          "flexShrink": 0,
          "left": "auto",
          "marginLeft": "0%",
          "maxWidth": "100%",
          "minHeight": 1,
          "paddingLeft": 15,
          "paddingRight": 15,
          "position": "relative",
          "right": "auto",
          "width": "100%",
        }
      }
    >
      <div
        className="search-box"
      >
        <img
          alt="search-icon"
          className="search-icon"
          src="/search-icon.svg"
        />
        <input
          className="input-search-keyword"
          data-testid="input-search-keyword"
          onChange={[Function]}
          placeholder="Search by author, tag or date"
          value=""
        />
      </div>
    </div>
    <div
      className="loading-result"
      data-testid="loading-result"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
      <div
        className="loading-text"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;

exports[`(Tags: RMCT-004) Flickr Search Page Tests Check Flickr search page details using snapshot 1`] = `
<div
  className="flickr-search-page"
>
  <div
    style={
      Object {
        "alignItems": "normal",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "flexShrink": 0,
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": -15,
        "marginRight": -15,
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "flexBasis": "100%",
          "flexGrow": 0,
          "flexShrink": 0,
          "left": "auto",
          "marginLeft": "0%",
          "maxWidth": "100%",
          "minHeight": 1,
          "paddingLeft": 15,
          "paddingRight": 15,
          "position": "relative",
          "right": "auto",
          "width": "100%",
        }
      }
    >
      <div
        className="search-box"
      >
        <img
          alt="search-icon"
          className="search-icon"
          src="/search-icon.svg"
        />
        <input
          className="input-search-keyword"
          data-testid="input-search-keyword"
          onChange={[Function]}
          placeholder="Search by author, tag or date"
          value=""
        />
      </div>
    </div>
    <div
      className="loading-result"
      data-testid="loading-result"
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 80,
            "width": 80,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
      <div
        className="loading-text"
      >
        Loading
      </div>
    </div>
  </div>
</div>
`;
